program/tests/deactivate.rs

```diff
@@ -0,0 +1,128 @@
+#![allow(clippy::arithmetic_side_effects)]
+
+mod helpers;
+
+use {
+    helpers::{context::StakeTestContext, lifecycle::StakeLifecycle},
+    mollusk_svm::result::Check,
+    solana_account::ReadableAccount,
+    solana_program_error::ProgramError,
+    solana_pubkey::Pubkey,
+    solana_stake_client::instructions::{DeactivateBuilder, DelegateStakeBuilder},
+    solana_stake_interface::{error::StakeError, state::StakeStateV2},
+    solana_stake_program::id,
+    test_case::test_case,
+};
+
+#[test_case(false; "activating")]
+#[test_case(true; "active")]
+fn test_deactivate(activate: bool) {
+    let mut ctx = StakeTestContext::with_delegation();
+    let min_delegation = ctx.minimum_delegation.unwrap();
+    let rent_exempt_reserve = ctx.rent_exempt_reserve;
+    let staker = ctx.staker;
+    let withdrawer = ctx.withdrawer;
+
+    let (stake, mut stake_account) = ctx
+        .stake_account(StakeLifecycle::Initialized)
+        .staked_amount(min_delegation)
+        .build();
+
+    let (vote, vote_account_data) = ctx.vote_account.clone().unwrap();
+
+    // Deactivating an undelegated account fails
+    ctx.checks(&[Check::err(ProgramError::InvalidAccountData)])
+        .test_missing_signers(false)
+        .execute(
+            DeactivateBuilder::new()
+                .stake(stake)
+                .stake_authority(staker)
+                .instruction(),
```

Comment:
If we had per-instruction builders, we could essentially could pull all of the work passing ctx defaults into the instructions to be internal to the builder. If we did that, I think we could make individual tests like 50% smaller. Maybe I am still enamored by the token-wrap pattern of having tests like this small:

```rust
#[test]
fn test_incorrect_escrow_owner() {
    let incorrect_escrow_owner = Pubkey::new_unique();
    WrapBuilder::default()
        .unwrapped_escrow_owner(incorrect_escrow_owner)
        .check(Check::err(TokenWrapError::EscrowOwnerMismatch.into()))
        .execute();
}
```